sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1351,6 +1351,28 @@ object RepartitionByExpression {
   }
 }
 
+/**
+ * This operator does not guarantee the output partitioning, because the partition number will be
+ * optimized by AQE.
+ */
+case class AdaptiveRepartition(
+    partitionExpressions: Seq[Expression],
+    child: LogicalPlan) extends UnaryNode {
+  override def maxRows: Option[Long] = child.maxRows
+  override def output: Seq[Attribute] = child.output
+
+  lazy val numPartitions: Int = conf.numShufflePartitions
+
+  def partitioning: Partitioning = if (partitionExpressions.nonEmpty) {
+    HashPartitioning(partitionExpressions, numPartitions)
```

Comment:
Does it 1. expand to that static number first and 2. attempt to coalesce with 3. falling back to the local shuffle readers?

Or does AQE has a logic to automatically figure out the number fo expand?

Reply:
> Or does AQE has a logic to automatically figure out the number fo expand?

yea, AQE can rebalance partitions automatically, can see PR [#32883](https://github.com/apache/spark/pull/32883)